pkg/ingester/ingester_v2.go

```diff
@@ -1438,6 +1466,22 @@ func (i *Ingester) numSeriesInTSDB() float64 {
 	return float64(count)
 }
 
+// getOldestUnshippedBlockMetric returns the unix timestamp of the oldest unshipped block or
+// 0 if all blocks have been shipped.
+func (i *Ingester) getOldestUnshippedBlockMetric() float64 {
+	i.userStatesMtx.RLock()
+	defer i.userStatesMtx.RUnlock()
+
+	oldest := uint64(0)
+	for _, db := range i.TSDBState.dbs {
+		if ts, err := db.getOldestUnshippedBlockTime(); err == nil && (oldest == 0 || ts < oldest) {
```

Comment:
Instead of computing this value on each metric collection, should we cache it and update after each call to the shipper? As it is now, each metric collection needs to read shipper json file for each tenant, which could be slow if there are many tenants.

Reply:
I can benchmark it and then we can decide if extra complexity is worth it. But agree checking it.

Reply:
I've written a simple benchmark ([commit](https://github.com/cortexproject/cortex/pull/3705/commits/9447e9a184064c4f1d91c846b80083034f28d7f5)) and on my MacBook it takes 31μs/op. If you have 1K tenants, it would be 31ms. I've the feeling caching for it, at this stage, would just be premature optimisation.

```
BenchmarkUserTSDB_getOldestUnshippedBlockTime-12    	   37930	     30898 ns/op	    4486 B/op	      26 allocs/op
```

Reply:
Is this a valid benchmark? On a busy server `getOldestUnshippedBlockTime` will not benefit from heavy caching of all IO operations as much as it does in this standalone test.

I don't think it would be a premature optimization, just a better approach. For 1K tenants, current approach needs to open/read/close 1000 files on each metric collection (=3000 syscalls). Instead we could simply read one atomic variable per tenant, and update it only when it can actually change. I don't see it as "premature optimization".

Reply:
+1, I share concern this could be pretty inefficient for large number of tenants.

Reply:
Ok, got it. I pushed a change to keep the shipped blocks in memory (they're tiny) and reload them each time any block is uploaded. I've also took the opportunity to improve the tests to not use a mocked shipper but the real one backed by a filesystem bucket.

What do you think?